TST: Fix TestArrayAlmostEqualNulp

rnelsonchem opened the pull request:
This fix addresses Issue #347.

All of the tests in this class were using random numbers generated without a
seed. In certain cases, this broke `test_simple` in a nondeterministic manner.
Part of the problem for this function is that `nulp` was being set to
2*max(random deviations). This value could be less than one depending on the
random numbers generated. Reset `nulp=1` which is the smallest value that
makes sense. Create a random array of deviations and force them to be in the
range [-1., 1.], because the abs(randn) can occasionally be > 1.

The assertion in test_big_float32 used a `nulp=1000`, which seemed excessive
for this particular test. Changed this to `nulp=1`, because that corresponds
to a value of ~1024 for this test case.

Documentation and random seeds were added to all of the other functions in
this class to make things more explicit.

rnelsonchem committed:
TST: Update assert_array_almost_equal_nulp tests

This fix addresses Issue #347.

All of the tests in this class were using random numbers generated without a
seed. This broke `test_simple` in a nondeterministic manner.

Rewrote all tests for this function. Instead of using random numbers,
well-defined arrays of widely varying amplitude were created instead. Changed
the array modification values so that the test arrays were either just above
or just below the NULP threshold.

Added tests for float64, float32, complex128, and complex64 dtypes, and added
both addition and subtraction tests. Tried to standardize the code for all
tests.

charris commented on numpy/testing/tests/test_utils.py:
```diff
@@ -468,48 +468,117 @@ def test_report_fail_percentage(self):
-            return assert_array_almost_equal_nulp(x, y,
-                                                  nulp=1000)
-        self.assertRaises(AssertionError, failure)
+        def failure(x, y):
```
The function definition would look better with blank lines before and after. Although I think you could dispense with the function and use `assert_array_almost_equal_nulp` directly.

charris commented on numpy/testing/tests/test_utils.py:
```diff
@@ -468,48 +468,117 @@ def test_report_fail_percentage(self):
         y = x + 1
-        def failure():
+        def failure(x, y):
             assert_array_almost_equal_nulp(x, y, nulp=1000)
```
See above.

charris commented:
Perhaps `linspace` or a list of carefully chosen numbers might be more useful than `randn`.

Generally looks good to me.

rnelsonchem commented:
@charris -- I agree with your assessment. I removed the old tests that used random numbers in favor of my new tests that just use `linspace`. This feels more in the spirit of the function being tested, in my opinion. 

I think this is ready to go. I can squash everything is there are no objections.

charris commented on numpy/testing/tests/test_utils.py:
```diff
@@ -468,47 +468,70 @@ def test_report_fail_percentage(self):
+        x = 10**x
+        x = np.r_[-x, x]
+        eps = np.finfo(x.dtype).eps
+        y = x + x*eps*nulp/2.
```
Maybe should check `y = x - x*eps*nulp/2` also.

rnelsonchem replied:
In this case, I think that [finfo](http://docs.scipy.org/doc/numpy-dev/reference/generated/numpy.finfo.html) also defines a epsneg for the subtraction case. Should I add new test functions for all those cases as well? (Not a problem, just want to make sure.)

charris replied:
Looks like epsneg is just half of eps. It probably makes sense to use that for the negative case. I wouldn't add new test functions, just add new tests in the existing functions, the traceback on failure will indicate the relevant line. Note that complex comes in both float32/float64 forms also.

charris commented on numpy/testing/tests/test_utils.py:
```diff
@@ -468,47 +468,70 @@ def test_report_fail_percentage(self):
+        eps = np.finfo(x.dtype).eps
+        y = x + x*eps*nulp*2.
+        self.assertRaises(AssertionError, assert_array_almost_equal_nulp, 
+                x, y, nulp)
```
PEP8, the second line should line up with `AssertionError` above. Also in the rest of the cases below.

charris commented on numpy/testing/tests/test_utils.py:
```diff
@@ -468,47 +468,70 @@ def test_report_fail_percentage(self):
-        x = np.random.randn(10)
-        y = np.array(x, np.complex) + 1e-16 * np.random.randn(10)
+    def test_float64_pass(self):
+        nulp = 5
```
Any particular reason to choose `nulp = 5`? If so, a comment would help. Could put things in a loop to check several values.

rnelsonchem replied:
I picked 5 because I wanted it to be a close to the precision limit, but I didn't want to be right at it it (nulp=1) in case that led to unforeseen problems. As it stands this is much tighter than the previous tests both in nulp and in the deviations needed to pass/fail the test. I can't think of a reason why we would need a loop. I'm happy to add one if you think it is needed.

charris commented:
Looks *much* nicer. Couple of comments and a PEP8 nitpick.

rnelsonchem commented:
Okay. I've added tests for subtraction with `epsneg` to all tests and also added complex64 tests as well. If this looks okay, I'll squash it into one commit.

charris commented on numpy/testing/tests/test_utils.py:
```diff
@@ -468,47 +468,150 @@ def test_report_fail_percentage(self):
+        xi = x + x*1j
+        yi = y + x*1j
+        assert_array_almost_equal_nulp(xi, yi, nulp)
+
```
Should probably check the imaginary parts also ;)

charris replied:
I think just `yi = y + y*1j` should be enough.

rnelsonchem commented:
@charris You are correct. I should have had the imaginary part change as well. I just changed that a couple of commits back, because I wanted the change in the original array to be equal to the floating point tests. I changed things back to `yi = y + y*1j` and just added a comment on the first instance.

I think this is now fairly complete.

njsmith commented on numpy/testing/tests/test_utils.py:
```diff
@@ -468,47 +468,152 @@ def test_report_fail_percentage(self):
+        # tests equivalent to non-complex ones
+        xi = x + x*1j
+        yi = y + y*1j
+        assert_array_almost_equal_nulp(xi, yi, nulp)
```
Sorry, I have one more nitpick... can we make this
```
assert_array_almost_equal_nulp(xi, y + x*1j, nulp)
assert_array_almost_equal_nulp(xi, x + y*1j, nulp)
assert_array_almost_equal_nulp(xi, y + y*1j, nulp)
```
and similarly below, so as to make sure that `assert_array_almost_equal_nulp` is actually looking at both the real and imaginary parts? The current test could pass e.g. if it didn't realize it was a complex array and was looking at the real parts only.

(Also I think the comment about "use modified `y` array only once..." may be out of date?)

charris replied:
Doesn't hurt, but I don't think that will make much difference because x is closer to x than y is ;)  The best place to check that both real and imaginary parts are looked at is in the tests for failure.

njsmith replied:
Hah, fair enough. I did mean this as a general comment on all the complex
tests of course :-)

On Sat, Jan 10, 2015 at 5:33 PM, Charles Harris <notifications@github.com>
wrote:

> In numpy/testing/tests/test_utils.py
> <https://github.com/numpy/numpy/pull/5347#discussion-diff-22760563>:
>
> > +
> > +    def test_complex128_pass(self):
> > +        nulp = 5
> > +        x = np.linspace(-20, 20, 50, dtype=np.float64)
> > +        x = 10**x
> > +        x = np.r_[-x, x]
> > +
> > +        eps = np.finfo(x.dtype).eps
> > +        # The test condition needs to be a factor of two lower compared to
> > +        # pure floats because real and imaginary parts will change
> > +        y = x + x*eps*nulp/4.
> > +        # Convert to complex, use modified `y` array only once in `yi` to keep
> > +        # tests equivalent to non-complex ones
> > +        xi = x + x*1j
> > +        yi = y + y*1j
> > +        assert_array_almost_equal_nulp(xi, yi, nulp)
>
> Doesn't hurt, but I don't think that will make much difference because x
> is closer to x than y is ;) The best place to check that both real and
> imaginary parts are looked at is in the tests for failure.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/numpy/numpy/pull/5347/files#r22760563>.
>



-- 
Nathaniel J. Smith
Postdoctoral researcher - Informatics - University of Edinburgh
http://vorpus.org

rnelsonchem commented:
Good points. Thanks. 

To make things more complete, I added tests for all of the complex number permutations. I removed the outdated comment (sorry that was sloppy), but I added a comment, though, for a couple sub-tests. When `y + y*1j`, the modification to the original array `x` needs to be half as much because both the real and imag parts are changing.

charris commented on numpy/testing/tests/test_utils.py:
```diff
@@ -468,47 +468,172 @@ def test_report_fail_percentage(self):
+        assert_array_almost_equal_nulp(xi, x + y*1j, nulp)
+        assert_array_almost_equal_nulp(xi, y + x*1j, nulp)
+        # The test condition needs to be a factor of two smaller because the
+        # real and imaginary parts both change
```
So the changes are taken together rather than separately?

rnelsonchem replied:
Yes. There is a `np.abs(xi-yi)` in the assert test, and with complex arrays where both parts are changing, it leads to a larger deviation than the other dtypes. I guess it is by a factor of sqrt(2), not 2. I'll change the comment. I could have made it exactly sqrt(2) difference, but I don't know that is necessary. I picked deviations that just work or just don't work. This is still pretty close.

charris commented:
LGTM, you can squash the commits now and we'll put them in.

rnelsonchem commented:
Okay. I've squashed it all into one commit. Seems to be passing all the test.

Thanks for the input and assistance everyone.

charris commented:
I cleaned up the commit message a bit in #5439, the line width needs to be 72 characters. You are still the author of the patch. Thanks.
